java/src/main/java/com/google/appengine/tools/pipeline/impl/backend/CloudTasksTaskQueue.java

```diff
@@ -0,0 +1,118 @@
+package com.google.appengine.tools.pipeline.impl.backend;
+
+import com.google.appengine.tools.pipeline.impl.servlets.TaskHandler;
+import com.google.appengine.tools.pipeline.impl.tasks.PipelineTask;
+
+import com.google.cloud.tasks.v2.*;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import lombok.AllArgsConstructor;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.google.appengine.tools.pipeline.impl.PipelineManager.DEFAULT_QUEUE_NAME;
+
+/**
+ * implementation of PipelineTaskQueue backed by Cloud Tasks
+ *
+ * TODO: what happens when tasks exists?
+ * TODO: retries for transients + internal errors
+ *
+ */
+@AllArgsConstructor(onConstructor_ = @Inject)
+public class CloudTasksTaskQueue implements PipelineTaskQueue {
+
+  AppEngineEnvironment appEngineEnvironment;
+
+  Provider<CloudTasksClient> cloudTasksClientProvider;
+
+  AppEngineServicesService appEngineServicesService;
+
+  @Override
+  public TaskReference enqueue(PipelineTask pipelineTask) {
+    return enqueue(List.of(pipelineTask)).iterator().next();
+  }
+
+  @Override
+  public TaskReference enqueue(String queueName, TaskSpec build) {
+    return enqueue(queueName, List.of(build)).iterator().next();
+  }
+
+  @Override
+  public Collection<TaskReference> enqueue(Collection<PipelineTask> pipelineTasks) {
+
+    Map<String, List<PipelineTask>> tasksByQueue = pipelineTasks.stream()
+      .collect(Collectors.groupingBy(pipelineTask -> Optional.ofNullable(pipelineTask.getQueueSettings().getOnQueue()).orElse(DEFAULT_QUEUE_NAME)));
+
+    /// probably could use parallelStream here, but in practice don't *really* expect to have multiple queues in the batch
+   return tasksByQueue.entrySet().stream()
+      .map(e -> enqueue(e.getKey(), e.getValue().stream()
+        .map(pipelineTask -> {
+          String host = appEngineServicesService.getWorkerServiceHostName(pipelineTask.getQueueSettings().getOnService(), pipelineTask.getQueueSettings().getOnServiceVersion());
+          return pipelineTask.toTaskSpec(host, TaskHandler.handleTaskUrl());
+        }).collect(Collectors.toList())))
+     .flatMap(Collection::stream)
+     .collect(Collectors.toList());
+  }
+
+  Collection<TaskReference> enqueue(String queue, Collection<TaskSpec> taskSpecs) {
+    QueueName queueName = QueueName.of(appEngineEnvironment.getProjectId(), appEngineEnvironment.getLocation(), queue);
+    try (CloudTasksClient cloudTasksClient = cloudTasksClientProvider.get()) {
+      return taskSpecs.parallelStream() //q: this safe? efficient?
```

Comment:
probably, but documentation doesn't specify thread-safety. 
I'd say the gain is minimal, so probably wouldn't bother just in case.